java/memory/src/main/java/org/apache/arrow/memory/LowCostIdentityHasMap.java

```diff
@@ -0,0 +1,338 @@
+  private static final int loadFactor = 7500;
+
+  /**
+   * Creates an Map with default expected maximum size.
+   */
+  public LowCostIdentityHasMap() {
+    this(DEFAULT_MIN_SIZE);
+  }
+
+  /**
+   * Creates an Map with the specified maximum size parameter.
+   *
+   * @param maxSize
+   *            The estimated maximum number of entries that will be put in
+   *            this map.
+   */
+  public LowCostIdentityHasMap(int maxSize) {
+    if (maxSize >= 0) {
+      this.size = 0;
+      threshold = getThreshold(maxSize);
+      elementData = newElementArray(computeElementArraySize());
+    } else {
+      throw new IllegalArgumentException();
+    }
+  }
+
+  private int getThreshold(int maxSize) {
+    // assign the threshold to maxSize initially, this will change to a
+    // higher value if rehashing occurs.
+    return maxSize > 2 ? maxSize : 2;
+  }
+
+  private int computeElementArraySize() {
+    int arraySize = (int) (((long) threshold * 10000) / loadFactor);
+    // ensure arraySize is positive, the above cast from long to int type
+    // leads to overflow and negative arraySize if threshold is too big
+    return arraySize < 0 ? -arraySize : arraySize;
+  }
+
+  /**
+   * Create a new element array
+   *
+   * @param s
+   *            the number of elements
+   * @return Reference to the element array
+   */
+  private Object[] newElementArray(int s) {
+    return new Object[s];
+  }
+
+  private V massageValue(Object value) {
+    return (V) ((value == null) ? null : value);
+  }
+
+  /**
+   * Removes all elements from this map, leaving it empty.
+   *
+   * @see #isEmpty()
+   * @see #size()
+   */
+  public void clear() {
+    size = 0;
+    for (int i = 0; i < elementData.length; i++) {
+      elementData[i] = null;
+    }
+  }
+
+  /**
+   * Returns whether this map contains the specified key.
+   *
+   * @param key
+   *            the key to search for.
+   * @return {@code true} if this map contains the specified key,
+   *         {@code false} otherwise.
+   */
+  public boolean containsKey(Object key) {
+    Preconditions.checkNotNull(key);
+
+    int index = findIndex(key, elementData);
+    return (elementData[index] == null) ? false : ((V)elementData[index]).getKey() == key;
+  }
+
+  /**
+   * Returns whether this map contains the specified value.
+   *
+   * @param value
+   *            the value to search for.
+   * @return {@code true} if this map contains the specified value,
+   *         {@code false} otherwise.
+   */
+  public boolean containsValue(V value) {
+    Preconditions.checkNotNull(value);
+
+    for (int i = 0; i < elementData.length; i++) {
+      if (elementData[i] == value) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Returns the value of the mapping with the specified key.
+   *
+   * @param key
+   *            the key.
+   * @return the value of the mapping with the specified key.
+   */
+  public V get(Object key) {
```

Comment:
same question as above

Reply:
The only thing that can be done is : public <K> V get(K key)
But in this case it is quite detached from "K" defined in 
ValueWithKeyIncluded<K>

Reply:
Hmmm - it does not like angle brackets in comments:
<code> 
public \<K\> V get(K key)
<code>